File: research/scicomp/Python/Crystallography/core/crystal_structure.py
"""
Crystal Structure Analysis
Comprehensive crystallographic structure representation and analysis tools.
Includes lattice parameter calculations, unit cell operations, and
crystallographic coordinate transformations.
"""
import numpy as np
from typing import List, Tuple, Dict, Optional, Union
from dataclasses import dataclass
import warnings
@dataclass
class LatticeParameters:
    """Crystal lattice parameters container."""
    a: float  # Lattice parameter a (Å)
    b: float  # Lattice parameter b (Å)
    c: float  # Lattice parameter c (Å)
    alpha: float  # Angle alpha (degrees)
    beta: float   # Angle beta (degrees)
    gamma: float  # Angle gamma (degrees)
    def __post_init__(self):
        """Validate lattice parameters."""
        if self.a <= 0 or self.b <= 0 or self.c <= 0:
            raise ValueError("Lattice parameters must be positive")
        if not (0 < self.alpha < 180 and 0 < self.beta < 180 and 0 < self.gamma < 180):
            raise ValueError("Lattice angles must be between 0 and 180 degrees")
    @property
    def alpha_rad(self) -> float:
        """Alpha angle in radians."""
        return np.radians(self.alpha)
    @property
    def beta_rad(self) -> float:
        """Beta angle in radians."""
        return np.radians(self.beta)
    @property
    def gamma_rad(self) -> float:
        """Gamma angle in radians."""
        return np.radians(self.gamma)
@dataclass
class AtomicPosition:
    """Atomic position in crystal structure."""
    element: str
    x: float  # Fractional coordinate x
    y: float  # Fractional coordinate y
    z: float  # Fractional coordinate z
    occupancy: float = 1.0
    thermal_factor: float = 0.0  # Isotropic B-factor (Å²)
    def __post_init__(self):
        """Validate atomic position parameters."""
        if not (0 <= self.occupancy <= 1):
            raise ValueError("Occupancy must be between 0 and 1")
        if self.thermal_factor < 0:
            raise ValueError("Thermal factor must be non-negative")
class CrystalStructure:
    """
    Crystal structure representation and analysis.
    Features:
    - Lattice parameter calculations
    - Unit cell volume and density
    - Coordinate transformations (fractional ↔ Cartesian)
    - Interatomic distance calculations
    - Miller indices operations
    - Structure factor calculations
    Examples:
        >>> lattice = LatticeParameters(a=5.0, b=5.0, c=5.0, alpha=90, beta=90, gamma=90)
        >>> atoms = [AtomicPosition('Si', 0.0, 0.0, 0.0), AtomicPosition('Si', 0.5, 0.5, 0.5)]
        >>> crystal = CrystalStructure(lattice, atoms)
        >>> print(f"Unit cell volume: {crystal.unit_cell_volume():.2f} Å³")
    """
    def __init__(self, lattice_parameters: LatticeParameters, atoms: List[AtomicPosition]):
        """
        Initialize crystal structure.
        Parameters:
            lattice_parameters: Crystal lattice parameters
            atoms: List of atomic positions
        """
        self.lattice = lattice_parameters
        self.atoms = atoms
        # Compute fundamental matrices
        self._metric_tensor = None
        self._direct_matrix = None
        self._reciprocal_matrix = None
        self._compute_matrices()
    def _compute_matrices(self):
        """Compute fundamental crystallographic matrices."""
        # Direct lattice matrix (columns are lattice vectors)
        a, b, c = self.lattice.a, self.lattice.b, self.lattice.c
        alpha, beta, gamma = self.lattice.alpha_rad, self.lattice.beta_rad, self.lattice.gamma_rad
        # Volume calculation
        cos_alpha, cos_beta, cos_gamma = np.cos(alpha), np.cos(beta), np.cos(gamma)
        sin_alpha, sin_beta, sin_gamma = np.sin(alpha), np.sin(beta), np.sin(gamma)
        volume = a * b * c * np.sqrt(1 - cos_alpha**2 - cos_beta**2 - cos_gamma**2
                                    + 2*cos_alpha*cos_beta*cos_gamma)
        # Direct lattice matrix
        self._direct_matrix = np.array([
            [a, b*cos_gamma, c*cos_beta],
            [0, b*sin_gamma, c*(cos_alpha - cos_beta*cos_gamma)/sin_gamma],
            [0, 0, volume/(a*b*sin_gamma)]
        ])
        # Metric tensor (dot products of lattice vectors)
        self._metric_tensor = np.array([
            [a**2, a*b*cos_gamma, a*c*cos_beta],
            [a*b*cos_gamma, b**2, b*c*cos_alpha],
            [a*c*cos_beta, b*c*cos_alpha, c**2]
        ])
        # Reciprocal lattice matrix
        self._reciprocal_matrix = 2*np.pi * np.linalg.inv(self._direct_matrix).T
    @property
    def direct_matrix(self) -> np.ndarray:
        """Direct lattice matrix (3x3)."""
        return self._direct_matrix.copy()
    @property
    def reciprocal_matrix(self) -> np.ndarray:
        """Reciprocal lattice matrix (3x3)."""
        return self._reciprocal_matrix.copy()
    @property
    def metric_tensor(self) -> np.ndarray:
        """Metric tensor (3x3)."""
        return self._metric_tensor.copy()
    def unit_cell_volume(self) -> float:
        """Calculate unit cell volume in Å³."""
        return np.abs(np.linalg.det(self._direct_matrix))
    def density(self, molecular_weight: float, z: int = 1) -> float:
        """
        Calculate crystal density.
        Parameters:
            molecular_weight: Molecular weight (g/mol)
            z: Number of formula units per unit cell
        Returns:
            Density in g/cm³
        """
        avogadro = 6.02214076e23  # mol⁻¹
        volume_cm3 = self.unit_cell_volume() * 1e-24  # Å³ to cm³
        return (z * molecular_weight) / (avogadro * volume_cm3)
    def fractional_to_cartesian(self, fractional_coords: np.ndarray) -> np.ndarray:
        """
        Convert fractional coordinates to Cartesian coordinates.
        Parameters:
            fractional_coords: Fractional coordinates (Nx3 or 3,)
        Returns:
            Cartesian coordinates in Å
        """
        fractional_coords = np.atleast_2d(fractional_coords)
        return (self._direct_matrix @ fractional_coords.T).T
    def cartesian_to_fractional(self, cartesian_coords: np.ndarray) -> np.ndarray:
        """
        Convert Cartesian coordinates to fractional coordinates.
        Parameters:
            cartesian_coords: Cartesian coordinates in Å (Nx3 or 3,)
        Returns:
            Fractional coordinates
        """
        cartesian_coords = np.atleast_2d(cartesian_coords)
        return (np.linalg.inv(self._direct_matrix) @ cartesian_coords.T).T
    def interatomic_distance(self, atom1_idx: int, atom2_idx: int,
                           include_symmetry: bool = False) -> float:
        """
        Calculate distance between two atoms.
        Parameters:
            atom1_idx: Index of first atom
            atom2_idx: Index of second atom
            include_symmetry: Whether to consider periodic boundary conditions
        Returns:
            Distance in Å
        """
        if atom1_idx >= len(self.atoms) or atom2_idx >= len(self.atoms):
            raise IndexError("Atom index out of range")
        atom1 = self.atoms[atom1_idx]
        atom2 = self.atoms[atom2_idx]
        # Fractional coordinate difference
        df = np.array([atom2.x - atom1.x, atom2.y - atom1.y, atom2.z - atom1.z])
        if include_symmetry:
            # Apply minimum image convention
            df = df - np.round(df)
        # Distance using metric tensor
        distance_squared = df @ self._metric_tensor @ df
        return np.sqrt(distance_squared)
    def d_spacing(self, h: int, k: int, l: int) -> float:
        """
        Calculate d-spacing for Miller indices (hkl).
        Parameters:
            h, k, l: Miller indices
        Returns:
            d-spacing in Å
        """
        # Reciprocal lattice vector
        hkl = np.array([h, k, l])
        # Calculate |G|² where G is reciprocal lattice vector
        reciprocal_metric = np.linalg.inv(self._metric_tensor)
        g_squared = hkl @ reciprocal_metric @ hkl
        if g_squared <= 0:
            raise ValueError("Invalid Miller indices result in zero d-spacing")
        return 1.0 / np.sqrt(g_squared)
    def bragg_angle(self, h: int, k: int, l: int, wavelength: float) -> float:
        """
        Calculate Bragg angle for reflection (hkl).
        Parameters:
            h, k, l: Miller indices
            wavelength: X-ray wavelength in Å
        Returns:
            Bragg angle in degrees
        """
        d = self.d_spacing(h, k, l)
        sin_theta = wavelength / (2 * d)
        if sin_theta > 1:
            raise ValueError(f"No diffraction possible for hkl=({h},{k},{l}) at λ={wavelength} Å")
        return np.degrees(np.arcsin(sin_theta))
    def systematic_absences(self, space_group: str) -> List[Tuple[int, int, int]]:
        """
        Determine systematic absences for given space group.
        Parameters:
            space_group: Space group symbol (e.g., 'Fm3m', 'P21/c')
        Returns:
            List of forbidden reflections
        """
        # Simplified implementation for common space groups
        forbidden = []
        # Face-centered cubic (F)
        if space_group.startswith('F'):
            # h+k, h+l, k+l must all be even
            for h in range(-5, 6):
                for k in range(-5, 6):
                    for l in range(-5, 6):
                        if (h + k) % 2 != 0 or (h + l) % 2 != 0 or (k + l) % 2 != 0:
                            forbidden.append((h, k, l))
        # Body-centered (I)
        elif space_group.startswith('I'):
            # h+k+l must be even
            for h in range(-5, 6):
                for k in range(-5, 6):
                    for l in range(-5, 6):
                        if (h + k + l) % 2 != 0:
                            forbidden.append((h, k, l))
        # Add more space group rules as needed
        return forbidden
    def coordination_number(self, atom_idx: int, cutoff_radius: float = 3.0) -> int:
        """
        Calculate coordination number for an atom.
        Parameters:
            atom_idx: Index of central atom
            cutoff_radius: Maximum distance for coordination (Å)
        Returns:
            Coordination number
        """
        if atom_idx >= len(self.atoms):
            raise IndexError("Atom index out of range")
        coordination = 0
        central_atom = self.atoms[atom_idx]
        # Check all other atoms
        for i, atom in enumerate(self.atoms):
            if i == atom_idx:
                continue
            distance = self.interatomic_distance(atom_idx, i, include_symmetry=True)
            if distance <= cutoff_radius:
                coordination += 1
        # Also check periodic images (simplified)
        for dx in [-1, 0, 1]:
            for dy in [-1, 0, 1]:
                for dz in [-1, 0, 1]:
                    if dx == 0 and dy == 0 and dz == 0:
                        continue
                    for i, atom in enumerate(self.atoms):
                        if i == atom_idx:
                            continue
                        # Translate atom position
                        translated_pos = np.array([atom.x + dx, atom.y + dy, atom.z + dz])
                        central_pos = np.array([central_atom.x, central_atom.y, central_atom.z])
                        df = translated_pos - central_pos
                        distance_squared = df @ self._metric_tensor @ df
                        distance = np.sqrt(distance_squared)
                        if distance <= cutoff_radius:
                            coordination += 1
        return coordination
    def powder_pattern_positions(self, wavelength: float, max_2theta: float = 90.0,
                                min_d_spacing: float = 1.0) -> List[Dict]:
        """
        Calculate powder diffraction peak positions.
        Parameters:
            wavelength: X-ray wavelength in Å
            max_2theta: Maximum 2θ angle in degrees
            min_d_spacing: Minimum d-spacing to consider (Å)
        Returns:
            List of reflection information dictionaries
        """
        reflections = []
        # Generate Miller indices up to reasonable limits
        max_h = max(5, int(2*self.lattice.a / min_d_spacing))
        max_k = max(5, int(2*self.lattice.b / min_d_spacing))
        max_l = max(5, int(2*self.lattice.c / min_d_spacing))
        for h in range(-max_h, max_h + 1):
            for k in range(-max_k, max_k + 1):
                for l in range(-max_l, max_l + 1):
                    if h == 0 and k == 0 and l == 0:
                        continue
                    try:
                        d = self.d_spacing(h, k, l)
                        if d < min_d_spacing:
                            continue
                        theta = self.bragg_angle(h, k, l, wavelength)
                        two_theta = 2 * theta
                        if two_theta <= max_2theta:
                            reflections.append({
                                'hkl': (h, k, l),
                                'd_spacing': d,
                                'theta': theta,
                                '2theta': two_theta,
                                'multiplicity': self._calculate_multiplicity(h, k, l)
                            })
                    except ValueError:
                        # Skip reflections that don't satisfy Bragg condition
                        continue
        # Sort by 2θ angle
        reflections.sort(key=lambda x: x['2theta'])
        return reflections
    def _calculate_multiplicity(self, h: int, k: int, l: int) -> int:
        """Calculate multiplicity factor for reflection (simplified)."""
        # This is a simplified calculation
        # Full implementation would require space group information
        unique_indices = len(set([abs(h), abs(k), abs(l)]))
        if unique_indices == 1:  # All indices equal (e.g., 111)
            return 8
        elif unique_indices == 2:  # Two indices equal (e.g., 110)
            return 12
        else:  # All indices different (e.g., 123)
            return 24
    def get_fractional_coordinates(self) -> np.ndarray:
        """Get all atomic fractional coordinates as array."""
        coords = np.zeros((len(self.atoms), 3))
        for i, atom in enumerate(self.atoms):
            coords[i] = [atom.x, atom.y, atom.z]
        return coords
    def get_cartesian_coordinates(self) -> np.ndarray:
        """Get all atomic Cartesian coordinates as array."""
        fractional = self.get_fractional_coordinates()
        return self.fractional_to_cartesian(fractional)
    def supercell(self, nx: int, ny: int, nz: int) -> 'CrystalStructure':
        """
        Create supercell structure.
        Parameters:
            nx, ny, nz: Supercell dimensions
        Returns:
            New CrystalStructure representing supercell
        """
        # New lattice parameters
        new_lattice = LatticeParameters(
            a=self.lattice.a * nx,
            b=self.lattice.b * ny,
            c=self.lattice.c * nz,
            alpha=self.lattice.alpha,
            beta=self.lattice.beta,
            gamma=self.lattice.gamma
        )
        # Replicate atoms
        new_atoms = []
        for i in range(nx):
            for j in range(ny):
                for k in range(nz):
                    for atom in self.atoms:
                        new_atom = AtomicPosition(
                            element=atom.element,
                            x=(atom.x + i) / nx,
                            y=(atom.y + j) / ny,
                            z=(atom.z + k) / nz,
                            occupancy=atom.occupancy,
                            thermal_factor=atom.thermal_factor
                        )
                        new_atoms.append(new_atom)
        return CrystalStructure(new_lattice, new_atoms)

File: research/scicomp/MATLAB/Crystallography/core/CrystalStructure.m
classdef CrystalStructure < handle
    % CrystalStructure - Crystal structure representation and analysis
    %
    % Features:
    % - Lattice parameter calculations
    % - Unit cell volume and density
    % - Coordinate transformations (fractional ↔ Cartesian)
    % - Interatomic distance calculations
    % - Miller indices operations
    % - Structure factor calculations
    %
    % Example:
    %   lattice = struct('a', 5.0, 'b', 5.0, 'c', 5.0, 'alpha', 90, 'beta', 90, 'gamma', 90);
    %   atoms = struct('element', {'Si', 'Si'}, 'x', {0.0, 0.5}, 'y', {0.0, 0.5}, 'z', {0.0, 0.5});
    %   crystal = CrystalStructure(lattice, atoms);
    %   volume = crystal.unitCellVolume();
    properties (Access = private)
        lattice_params      % Lattice parameters structure
        atoms_list         % Cell array of atomic positions
        direct_matrix      % Direct lattice matrix (3x3)
        reciprocal_matrix  % Reciprocal lattice matrix (3x3)
        metric_tensor      % Metric tensor (3x3)
    end
    properties (Dependent)
        lattice            % Lattice parameters (read-only)
        atoms             % Atomic positions (read-only)
        directMatrix      % Direct lattice matrix (read-only)
        reciprocalMatrix  % Reciprocal lattice matrix (read-only)
        metricTensor      % Metric tensor (read-only)
    end
    methods
        function obj = CrystalStructure(lattice_parameters, atomic_positions)
            % Constructor
            %
            % Parameters:
            %   lattice_parameters - Structure with fields: a, b, c, alpha, beta, gamma
            %   atomic_positions - Structure array with fields: element, x, y, z, occupancy, thermal_factor
            % Validate inputs
            obj.validateLatticeParameters(lattice_parameters);
            obj.validateAtomicPositions(atomic_positions);
            obj.lattice_params = lattice_parameters;
            obj.atoms_list = atomic_positions;
            % Compute fundamental matrices
            obj.computeMatrices();
        end
        % Dependent property getters
        function lattice = get.lattice(obj)
            lattice = obj.lattice_params;
        end
        function atoms = get.atoms(obj)
            atoms = obj.atoms_list;
        end
        function matrix = get.directMatrix(obj)
            matrix = obj.direct_matrix;
        end
        function matrix = get.reciprocalMatrix(obj)
            matrix = obj.reciprocal_matrix;
        end
        function tensor = get.metricTensor(obj)
            tensor = obj.metric_tensor;
        end
        function volume = unitCellVolume(obj)
            % Calculate unit cell volume in Å³
            volume = abs(det(obj.direct_matrix));
        end
        function rho = density(obj, molecular_weight, z)
            % Calculate crystal density
            %
            % Parameters:
            %   molecular_weight - Molecular weight (g/mol)
            %   z - Number of formula units per unit cell (default: 1)
            %
            % Returns:
            %   rho - Density in g/cm³
            if nargin < 3
                z = 1;
            end
            validateattributes(molecular_weight, {'numeric'}, {'scalar', 'positive'});
            validateattributes(z, {'numeric'}, {'scalar', 'positive', 'integer'});
            avogadro = 6.02214076e23;  % mol⁻¹
            volume_cm3 = obj.unitCellVolume() * 1e-24;  % Å³ to cm³
            rho = (z * molecular_weight) / (avogadro * volume_cm3);
        end
        function cartesian_coords = fractionalToCartesian(obj, fractional_coords)
            % Convert fractional coordinates to Cartesian coordinates
            %
            % Parameters:
            %   fractional_coords - Fractional coordinates (Nx3 or 3x1)
            %
            % Returns:
            %   cartesian_coords - Cartesian coordinates in Å
            if size(fractional_coords, 2) == 3
                % Nx3 format
                cartesian_coords = (obj.direct_matrix * fractional_coords')';
            else
                % 3x1 or 3xN format
                cartesian_coords = obj.direct_matrix * fractional_coords;
            end
        end
        function fractional_coords = cartesianToFractional(obj, cartesian_coords)
            % Convert Cartesian coordinates to fractional coordinates
            %
            % Parameters:
            %   cartesian_coords - Cartesian coordinates in Å (Nx3 or 3x1)
            %
            % Returns:
            %   fractional_coords - Fractional coordinates
            inv_direct = inv(obj.direct_matrix);
            if size(cartesian_coords, 2) == 3
                % Nx3 format
                fractional_coords = (inv_direct * cartesian_coords')';
            else
                % 3x1 or 3xN format
                fractional_coords = inv_direct * cartesian_coords;
            end
        end
        function distance = interatomicDistance(obj, atom1_idx, atom2_idx, include_symmetry)
            % Calculate distance between two atoms
            %
            % Parameters:
            %   atom1_idx - Index of first atom
            %   atom2_idx - Index of second atom
            %   include_symmetry - Whether to consider periodic boundary conditions (default: false)
            %
            % Returns:
            %   distance - Distance in Å
            if nargin < 4
                include_symmetry = false;
            end
            validateattributes(atom1_idx, {'numeric'}, {'scalar', 'integer', 'positive'});
            validateattributes(atom2_idx, {'numeric'}, {'scalar', 'integer', 'positive'});
            if atom1_idx > length(obj.atoms_list) || atom2_idx > length(obj.atoms_list)
                error('Atom index out of range');
            end
            atom1 = obj.atoms_list(atom1_idx);
            atom2 = obj.atoms_list(atom2_idx);
            % Fractional coordinate difference
            df = [atom2.x - atom1.x; atom2.y - atom1.y; atom2.z - atom1.z];
            if include_symmetry
                % Apply minimum image convention
                df = df - round(df);
            end
            % Distance using metric tensor
            distance_squared = df' * obj.metric_tensor * df;
            distance = sqrt(distance_squared);
        end
        function d = dSpacing(obj, h, k, l)
            % Calculate d-spacing for Miller indices (hkl)
            %
            % Parameters:
            %   h, k, l - Miller indices
            %
            % Returns:
            %   d - d-spacing in Å
            validateattributes(h, {'numeric'}, {'scalar', 'integer'});
            validateattributes(k, {'numeric'}, {'scalar', 'integer'});
            validateattributes(l, {'numeric'}, {'scalar', 'integer'});
            % Reciprocal lattice vector
            hkl = [h; k; l];
            % Calculate |G|² where G is reciprocal lattice vector
            reciprocal_metric = inv(obj.metric_tensor);
            g_squared = hkl' * reciprocal_metric * hkl;
            if g_squared <= 0
                error('Invalid Miller indices result in zero d-spacing');
            end
            d = 1.0 / sqrt(g_squared);
        end
        function theta = braggAngle(obj, h, k, l, wavelength)
            % Calculate Bragg angle for reflection (hkl)
            %
            % Parameters:
            %   h, k, l - Miller indices
            %   wavelength - X-ray wavelength in Å
            %
            % Returns:
            %   theta - Bragg angle in degrees
            d = obj.dSpacing(h, k, l);
            sin_theta = wavelength / (2 * d);
            if sin_theta > 1
                error('No diffraction possible for hkl=(%d,%d,%d) at λ=%g Å', h, k, l, wavelength);
            end
            theta = asind(sin_theta);
        end
        function coord_num = coordinationNumber(obj, atom_idx, cutoff_radius)
            % Calculate coordination number for an atom
            %
            % Parameters:
            %   atom_idx - Index of central atom
            %   cutoff_radius - Maximum distance for coordination (Å) (default: 3.0)
            %
            % Returns:
            %   coord_num - Coordination number
            if nargin < 3
                cutoff_radius = 3.0;
            end
            if atom_idx > length(obj.atoms_list)
                error('Atom index out of range');
            end
            coord_num = 0;
            % Check all other atoms
            for i = 1:length(obj.atoms_list)
                if i == atom_idx
                    continue;
                end
                distance = obj.interatomicDistance(atom_idx, i, true);
                if distance <= cutoff_radius
                    coord_num = coord_num + 1;
                end
            end
        end
        function supercell_crystal = supercell(obj, nx, ny, nz)
            % Create supercell structure
            %
            % Parameters:
            %   nx, ny, nz - Supercell dimensions
            %
            % Returns:
            %   supercell_crystal - New CrystalStructure representing supercell
            validateattributes(nx, {'numeric'}, {'scalar', 'integer', 'positive'});
            validateattributes(ny, {'numeric'}, {'scalar', 'integer', 'positive'});
            validateattributes(nz, {'numeric'}, {'scalar', 'integer', 'positive'});
            % New lattice parameters
            new_lattice = obj.lattice_params;
            new_lattice.a = new_lattice.a * nx;
            new_lattice.b = new_lattice.b * ny;
            new_lattice.c = new_lattice.c * nz;
            % Replicate atoms
            new_atoms = [];
            atom_count = 0;
            for i = 0:nx-1
                for j = 0:ny-1
                    for k = 0:nz-1
                        for atom_idx = 1:length(obj.atoms_list)
                            atom = obj.atoms_list(atom_idx);
                            atom_count = atom_count + 1;
                            new_atom.element = atom.element;
                            new_atom.x = (atom.x + i) / nx;
                            new_atom.y = (atom.y + j) / ny;
                            new_atom.z = (atom.z + k) / nz;
                            if isfield(atom, 'occupancy')
                                new_atom.occupancy = atom.occupancy;
                            else
                                new_atom.occupancy = 1.0;
                            end
                            if isfield(atom, 'thermal_factor')
                                new_atom.thermal_factor = atom.thermal_factor;
                            else
                                new_atom.thermal_factor = 0.0;
                            end
                            new_atoms = [new_atoms, new_atom];
                        end
                    end
                end
            end
            supercell_crystal = CrystalStructure(new_lattice, new_atoms);
        end
    end
    methods (Access = private)
        function validateLatticeParameters(~, lattice)
            % Validate lattice parameters structure
            required_fields = {'a', 'b', 'c', 'alpha', 'beta', 'gamma'};
            for i = 1:length(required_fields)
                if ~isfield(lattice, required_fields{i})
                    error('Lattice parameters must contain field: %s', required_fields{i});
                end
                value = lattice.(required_fields{i});
                if ~isnumeric(value) || ~isscalar(value)
                    error('Lattice parameter %s must be a numeric scalar', required_fields{i});
                end
                if strcmp(required_fields{i}(1), 'a') || strcmp(required_fields{i}(1), 'b') || strcmp(required_fields{i}(1), 'c')
                    if value <= 0
                        error('Lattice parameters a, b, c must be positive');
                    end
                else
                    if value <= 0 || value >= 180
                        error('Lattice angles must be between 0 and 180 degrees');
                    end
                end
            end
        end
        function validateAtomicPositions(~, atoms)
            % Validate atomic positions structure array
            if isempty(atoms)
                error('At least one atom must be provided');
            end
            required_fields = {'element', 'x', 'y', 'z'};
            for atom_idx = 1:length(atoms)
                atom = atoms(atom_idx);
                for i = 1:length(required_fields)
                    if ~isfield(atom, required_fields{i})
                        error('Atom %d must contain field: %s', atom_idx, required_fields{i});
                    end
                end
                if ~ischar(atom.element) && ~isstring(atom.element)
                    error('Atom %d element must be a string', atom_idx);
                end
                coords = [atom.x, atom.y, atom.z];
                if ~isnumeric(coords) || length(coords) ~= 3
                    error('Atom %d coordinates must be numeric values', atom_idx);
                end
                % Set default values for optional fields
                if ~isfield(atom, 'occupancy')
                    atoms(atom_idx).occupancy = 1.0;
                end
                if ~isfield(atom, 'thermal_factor')
                    atoms(atom_idx).thermal_factor = 0.0;
                end
            end
        end
        function computeMatrices(obj)
            % Compute fundamental crystallographic matrices
            a = obj.lattice_params.a;
            b = obj.lattice_params.b;
            c = obj.lattice_params.c;
            alpha = deg2rad(obj.lattice_params.alpha);
            beta = deg2rad(obj.lattice_params.beta);
            gamma = deg2rad(obj.lattice_params.gamma);
            % Trigonometric values
            cos_alpha = cos(alpha);
            cos_beta = cos(beta);
            cos_gamma = cos(gamma);
            sin_alpha = sin(alpha);
            sin_beta = sin(beta);
            sin_gamma = sin(gamma);
            % Volume calculation
            volume = a * b * c * sqrt(1 - cos_alpha^2 - cos_beta^2 - cos_gamma^2 + ...
                                     2*cos_alpha*cos_beta*cos_gamma);
            % Direct lattice matrix
            obj.direct_matrix = [
                a, b*cos_gamma, c*cos_beta;
                0, b*sin_gamma, c*(cos_alpha - cos_beta*cos_gamma)/sin_gamma;
                0, 0, volume/(a*b*sin_gamma)
            ];
            % Metric tensor (dot products of lattice vectors)
            obj.metric_tensor = [
                a^2, a*b*cos_gamma, a*c*cos_beta;
                a*b*cos_gamma, b^2, b*c*cos_alpha;
                a*c*cos_beta, b*c*cos_alpha, c^2
            ];
            % Reciprocal lattice matrix
            obj.reciprocal_matrix = 2*pi * inv(obj.direct_matrix)';
        end
    end
end